### PR TITLE
Option to preserve bitwise accuracy of gradient checkpointed vs non-checkpointed dropout

### PR DESCRIPTION
**Pull request by mcarilli**
This issue was noticed, and fix proposed, by @raulpuric.

Checkpointing is implemented by rerunning a forward-pass segment for each checkpointed segment during backward.  This can result in the RNG state advancing more than it would without checkpointing, which can cause checkpoints that include dropout invocations to lose end-to-end bitwise accuracy as compared to non-checkpointed passes.  

The present PR contains optional logic to juggle the RNG states such that checkpointed passes containing dropout achieve bitwise accuracy with non-checkpointed equivalents.**  The user requests this behavior by supplying `preserve_rng_state=True` to `torch.utils.checkpoint` or `torch.utils.checkpoint_sequential`.

Currently, `preserve_rng_state=True` may incur a moderate performance hit because restoring MTGP states can be expensive.  However, restoring Philox states is dirt cheap, so @syed-ahmed's [RNG refactor](https://github.com/pytorch/pytorch/pull/13070#discussion_r235179882), once merged, will make this option more or less free.

I'm a little wary of the [def checkpoint(function, *args, preserve_rng_state=False):](https://github.com/pytorch/pytorch/pull/14253/files#diff-58da227fc9b1d56752b7dfad90428fe0R75) argument-passing method (specifically, putting a kwarg after a variable argument list).  Python 3 seems happy with it.
Edit:  It appears Python 2.7 is NOT happy with a [kwarg after *args](https://travis-ci.org/pytorch/pytorch/builds/457706518?utm_source=github_status&utm_medium=notification).  `preserve_rng_state` also needs to be communicated in a way that doesn't break any existing usage.  I'm open to suggestions (a global flag perhaps)?

**Batchnorm may still be an issue, but that's a battle for another day.

### CODE REVIEW

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -24,9 +24,22 @@ def check_backward_validity(inputs):
...
+            # to the device during their run_fn.  Therefore, we stash both
+            # the cpu and cuda rng states unconditionally.
+            #
+            # TODO:
+            # We also can't know if the run_fn will internally move some args to a device
+            # other than the current device, which would require logic to preserve
+            # rng states for those devices as well...but I see no easy way to
+            # handle such cases.
+            ctx.fwd_cpu_rng_state = torch.get_rng_state()
+            ctx.fwd_cuda_rng_state = torch.cuda.get_rng_state()
```
Comment (soumith) [inline torch/utils/checkpoint.py]: this assumes user is using CUDA function. guard this with whether cuda lazy initialization has happened or not... (on current device)

File: torch/utils/checkpoint.py [hunk 2]
```diff
@@ -79,21 +104,34 @@ def checkpoint(function, *args):
         grads are needed for model inputs, otherwise the checkpointed part of the
         model won't have gradients.
 
+    .. warning:
+        Checkpointing is implemented by rerunning a forward-pass segment for
+        each checkpointed segment during backward.  This can result in running
+        states like the RNG state used for dropout to be advanced more than
+        they would be without checkpointing, which can cause checkpoints that
+        include dropout invocations to lose end-to-end bitwise accuracy as
```
Comment (soumith) [inline torch/utils/checkpoint.py]: "to lose end-to-end bitwise accuracy as
        compared to non-checkpointed passes"

-> "to have non-deterministic output compared to"

**Review by soumith — CHANGES_REQUESTED** [review]

File: torch/utils/checkpoint.py (hunk 1 above)
Comment (mcarilli) [reply torch/utils/checkpoint.py]: Oops, of course, I don't want to trigger initialization of a cuda context if the user doesn't require one.  There remains the (somewhat remote) possibility that the user intends cuda initialization to happen later (within their run_fn).  There's no way for me to anticipate that will happen, though...

Comment (mcarilli) [reply torch/utils/checkpoint.py]: Hopefully resolved by the latest commit

**Review by soumith — APPROVED** [review]
lgtm, let's see if tests pass

File: torch/utils/checkpoint.py [hunk 3]
```diff
@@ -37,17 +57,32 @@ def backward(ctx, *args):
...
+        if ctx.preserve_rng_state:
+            # Stash the surrounding rng state, and mimic the state that was
+            # present at this time during forward.  If cuda was not initialized
+            # at this time during the original forward, we assume that the imminent
+            # forward-segment-within-backward will not alter the cuda rng state.
+            current_cpu_rng_state = torch.get_rng_state()
+            torch.set_rng_state(ctx.fwd_cpu_rng_state)
+            if ctx.had_cuda_in_fwd:
+                current_cuda_rng_state = torch.cuda.get_rng_state()
+                torch.cuda.set_rng_state(ctx.fwd_cuda_rng_state)
```
Comment (apaszke) [inline torch/utils/checkpoint.py]: Can we implement a helper like `get_rng_states()` and `set_rng_states()` that deals with CUDA and other things internally? It could be a regular Python class that encapsulates all the logic. Also, it would be good to use context managers, so that we do restore the correct thing when an exception is thrown.

File: torch/utils/checkpoint.py [hunk 4]
```diff
@@ -37,17 +57,32 @@ def backward(ctx, *args):
...
 
         if isinstance(outputs, torch.Tensor):
             outputs = (outputs,)
         torch.autograd.backward(outputs, args)
-        return (None,) + tuple(inp.grad for inp in detached_inputs)
+        return (None, None) + tuple(inp.grad for inp in detached_inputs)
 
 
-def checkpoint(function, *args):
+def checkpoint(function, *args, preserve_rng_state=False):
```
Comment (apaszke) [inline torch/utils/checkpoint.py]: We should really default to `True` :confused:

File: torch/utils/checkpoint.py (hunk 3 above)
Comment (mcarilli) [reply torch/utils/checkpoint.py]: Do you want me to do that as part of this PR?

File: torch/utils/checkpoint.py (hunk 4 above)
Comment (mcarilli) [reply torch/utils/checkpoint.py]: I've moved `preserve_rng_state` to a global flag, defaulted to True.  This feels cleaner because we don't need to pass the additional arg around (which python 2.7 wasn't allowing anyway).

File: torch/utils/checkpoint.py (hunk 3 above)
Comment (syed-ahmed) [reply torch/utils/checkpoint.py]: Not sure how well this would fit but may be you could use `with torch.random.fork_rng():`? This exists in `torch/random.py`

Comment (mcarilli) [reply torch/utils/checkpoint.py]: @syed-ahmed Good find, this looks perfect.

**Review by apaszke — CHANGES_REQUESTED** [review]
I realized there are some pretty fundamental design flaws in what you're doing (we might not be caching the right RNG state). Also, please don't use global state unless it's necessary...

**Comment by mcarilli** [comment]
@apaszke I know we might not be preserving the right RNG states.  I left some comments to that effect in the code.  It's challenging because A. we can't know what devices will be used in the `run_fn`, and B. we also can't know if the user intends to initialize the cuda context on one or more devices within their `run_fn`.

To deal with A. I believe we have two choices:
1.  Stash and restore the RNG states for ALL devices visible to this process.  Probably overly conservative/paranoid for the majority of cases.
2.  Stash and restore the RNG states for the current device.
I went with option 2 because it seemed like a reasonable compromise.  I can add a warning to the docs that if the `run_fn` moves operations to other devices, bitwise accuracy is no longer guaranteed.

I have no idea how to deal with B., but B. seems like more of an edge case. 

The global state was necessary (imo) because I couldn't figure out a way to mix *args with kwargs in a way that was both python 2.7-compliant and would not break existing code.  I was prepared to experiment with other approaches but it appears you can live with the global flag.  `torch.backends.cudnn.deterministic` is also exposed as a global flag so this seemed symmetric to me.

**Comment by apaszke** [comment]
Some commentary:
1. Yes, I agree this is definitely too much, especially considering the overhead it adds.
2. I understand, but this really is a *bad compromise*. Most user PyTorch programs hardly ever change the current device, but simply specify `device=...` or use `.to(...)` to transfer some inputs to a different one, and then do operations on those tensors. This means that if someone uses more than a single device, you'll have a miss rate of almost 100%.

I think a better compromise would be to not use the current device to guide the heuristic, but *any tensor input*. That would be a reasonable solution I think, but it would still warrant some commentary in the docs.

---

Regarding your second issue, you should add `**kwargs` to the functions you're interested in, and then do something like this:
```py
def my_fn(arg1, *args, **kwargs):
    my_kwarg = kwargs.pop('my_kwarg', <default value>)
    if kwargs:
        raise ValueError("Unexpected keyword arguments: " + ",".join(arg for arg in kwargs))
```
